src/docfx/build/legacy/metadata/LegacyMetadata.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+using System.Text;
+using System.Web;
+
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class LegacyMetadata
+    {
+        private static readonly string[] s_pageMetadataOutputItems =
+       {
+            "author", "breadcrumb_path", "depot_name", "description", "document_id",
+            "document_version_independent_id", "gitcommit", "keywords",
+            "locale", "ms.author", "ms.date", "ms.prod", "ms.topic", "original_content_git_url",
+            "page_type", "search.ms_docsetname", "search.ms_product", "search.ms_sitename", "site_name",
+            "toc_rel", "uhfHeaderId", "updated_at", "version", "word_count",
+        };
+
+        private static readonly string[] s_metadataOutputItems =
+        {
+            "author", "breadcrumb_path", "canonical_url", "content_git_url", "depot_name", "description", "document_id",
+            "document_version_independent_id", "experiment_id", "experimental", "gitcommit", "is_dynamic_rendering", "keywords",
+            "layout", "locale", "ms.author", "ms.date", "ms.prod", "ms.topic", "open_to_public_contributors", "original_content_git_url",
+            "page_type", "search.ms_docsetname", "search.ms_product", "search.ms_sitename", "site_name", "title", "titleSuffix", "toc_asset_id",
+            "toc_rel", "uhfHeaderId", "updated_at", "version", "word_count", "redirect_url", "redirect_document_id",
+        };
+
+        public static JObject GenerateLegacyRawMetadata(PageModel pageModel, Docset docset, Document file, GitRepoInfoProvider repo, TableOfContentsMap tocMap)
+        {
+            var rawMetadata = pageModel.Metadata != null ? new JObject(pageModel.Metadata) : new JObject();
+            rawMetadata["fileRelativePath"] = Path.GetFileNameWithoutExtension(file.OutputPath) + ".html";
+            rawMetadata["toc_rel"] = pageModel.TocRelativePath ?? tocMap.FindTocRelativePath(file);
```

Comment:
which file does not have TocRelativePath?

Reply:
redirection document doesn't have toc relative path in our ideal output

---
In reply to: [196705003](https://github.com/dotnet/docfx/pull/2906#discussion_r196705003) [](ancestors = 196705003)